llvm/include/llvm/Config/abi-breaking.h.cmake

```diff
@@ -12,12 +12,36 @@
 #ifndef LLVM_ABI_BREAKING_CHECKS_H
 #define LLVM_ABI_BREAKING_CHECKS_H
 
+// llvm-config.h is required for LLVM_ENABLE_LLVM_EXPORT_ANNOTATIONS
+#include "llvm/Config/llvm-config.h"
+
 /* Define to enable checks that alter the LLVM C++ ABI */
 #cmakedefine01 LLVM_ENABLE_ABI_BREAKING_CHECKS
 
 /* Define to enable reverse iteration of unordered llvm containers */
 #cmakedefine01 LLVM_ENABLE_REVERSE_ITERATION
 
+// Properly annotate EnableABIBreakingChecks or DisableABIBreakingChecks for
+// export from shared library.
+#if !defined(LLVM_ABI_GENERATING_ANNOTATIONS)
+// TODO(https://github.com/llvm/llvm-project/issues/145406): eliminate need for
+// two preprocessor definitions to gate LLVM_ABI macro definitions.
+#if defined(LLVM_ENABLE_LLVM_EXPORT_ANNOTATIONS) && !defined(LLVM_BUILD_STATIC)
+#if defined(_WIN32) && !defined(__MINGW32__)
```

Comment:
IIRC AIX also uses `__declspec(dllimport)` and `__declspec(dllexport)` due to the GOFF object file format. Additionally, PS4/PS5 also use `__declspec(dllimport)` and `__declspec(dllexport)` even though use the ELF as the object file format.

I think that we can move forward with this for the time being, but this needs revision. We will need this to look like what I had sketched out originally with explicit cases for the OS. The generalised `__has_attribute(visibility)` just doesn't work IMO.